src/services/BlockchainScanner.ts

```diff
@@ -3,66 +3,112 @@ import {getModelForClass} from '@typegoose/typegoose';
 import {BlockchainMarker} from '../models/BlockchainMarker.js';
 import {Logger} from 'winston';
 import {StaticJsonRpcProvider} from '@ethersproject/providers';
+import {ChainsIds} from '../types/ChainsIds.js';
+import Settings from '../types/Settings.js';
+import {DexProtocolName} from '../types/DexProtocolName.js';
+import logger from '../lib/logger.js';
 
 export type Result = {
   fromBlock: number;
   toBlock?: number;
   success: boolean;
   synchronized?: boolean;
+  waitBlockRange?: boolean;
 };
 
 @injectable()
 export abstract class BlockchainScanner {
   @inject('Logger') logger!: Logger;
 
-  startBlock!: number;
-  blockchainId!: string;
-  markerId!: string;
-  step!: number;
+  startBlock: number;
+  dexProtocolName!: DexProtocolName;
+  step: number;
   provider: StaticJsonRpcProvider | undefined;
+  chainId!: ChainsIds;
+
+  constructor(chainId: ChainsIds, dexProtocolName: DexProtocolName, settings: Settings) {
+    this.step = 0;
+    this.logger = logger;
+
+    if (!chainId || !dexProtocolName) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] chainId or dexProtocolName not set`);
+    }
+
+    this.chainId = chainId;
+    this.dexProtocolName = dexProtocolName;
+
+    if (!settings.blockchains[chainId].providerUrl.join('')) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] provider URL not set`);
+    }
+
+    const startBlock = settings.dexes[chainId]?.[dexProtocolName]?.startBlock;
+
+    if (!startBlock || Number(startBlock) <= 0) {
+      throw new Error(`[BlockchainScanner][${chainId}][${dexProtocolName}] startBlock must be higher than 0`);
+    }
+
+    this.startBlock = startBlock!;
+    this.step = settings.dexes[chainId]?.[dexProtocolName]?.agentStep || 0;
+  }
 
   abstract apply(fromBlock: number, toBlock: number): Promise<boolean>;
 
   async run(): Promise<Result> {
     const fromBlock = await this.getMarker();
 
     if (!this.provider) {
-      this.logger.info('[BlockchainScanner] not active');
+      throw new Error(`[BlockchainScanner][${this.chainId}][${this.dexProtocolName}] getToBlock: no provider`);
+    }
+
+    const latestBlock = await this.provider.getBlockNumber();
+    const stepBlock = fromBlock + this.step;
+    const blocksToScan = latestBlock - fromBlock;
+
+    if (blocksToScan < this.step) {
+      this.logger.debug(
+        // eslint-disable-next-line max-len
+        `[BlockchainScanner][${this.chainId}][${this.dexProtocolName}] fromBlock: ${fromBlock}, latestBlock: ${latestBlock}, step: ${this.step}, stepBlock: ${stepBlock}`,
+      );
+      return {fromBlock, success: false, waitBlockRange: true};
+    }
+
+    const toBlock = await this.getToBlock(fromBlock, latestBlock);
+
+    if (!toBlock) {
```

Comment:
there is not way to not have toBlock at this point

Reply:
You are right, done!